Generate thumbnails

benoitblanc opened the pull request:
As I described in issue #20 , I try to generate thumbnails when it doesn't exist in directory.

I used code in QWC2 repository, with functions in `themesConfig.py`.

I deployed it on my QWC2 application and it works, but maybe there are some improvements to do.

Thanks,

Benoît

Close #20

HusseinKabbout commented on config_generator/map_viewer_config.py:
```diff
@@ -452,6 +459,132 @@ def theme_item(self, cfg_item):
+            try:
+                os.makedirs(self.qwc_base_dir + "/assets/img/genmapthumbs/")
+            except Exception as e:
+                if not isinstance(e, FileExistsError): raise e
```
Wouldn't it be better here if we just let the user know that the `genmapthumbs` directory couldn't be created instead of raising an exception?

For Example something like this:
```
except Exception as e:
    if not isinstance(e, FileExistsError):
        self.logger.error("The directory for auto generated thumbnails could not be created\n %s" % (str(e)))

HusseinKabbout commented:
@benoitblanc Thank you for your contribution! I added some comments to your PR.

benoitblanc commented:
Hello @HusseinKabbout , thanks for your review.

I have just added new commit to take your comments into account. Now I use `self.capabilites_reader` to get capabilities for projects so I deleted all the functions I added in first commit. Also, I added `CRS` into `capabilites_reader` because we need it to do `GetMap` request.

manisandro commented on config_generator/map_viewer_config.py:
```diff
@@ -73,6 +76,11 @@ def __init__(self, tenant_path, generator_config, capabilities_reader,
+        self.qwc_base_dir = generator_config.get('qwc2_base_dir')
+
+        self.base_url = "http://" + socket.gethostname()
+
```
Please use

    self.default_qgis_server_url = generator_config.get(
            'default_qgis_server_url', 'http://localhost:8001/ows/'
        ).rstrip('/') + '/'

as done in `capabilities_reader.py` and use that instead of `self.base_url` below.

manisandro commented on config_generator/map_viewer_config.py:
```diff
@@ -452,6 +458,71 @@ def theme_item(self, cfg_item):
+
+        item = self.capabilities_reader.service_name(cfg_item['url'])
+
+        capabilities = self.capabilities_reader.wms_capabilities[item]['root_layer']
```
You could pass the pre-existing `cap` and `service_name` variables read in `theme_item()` to `get_thumbnail()`, and then use `root_layer = cap.get('root_layer', {})` for consistency.

manisandro commented on config_generator/map_viewer_config.py:
```diff
@@ -452,6 +458,71 @@ def theme_item(self, cfg_item):
+
+        try:
+            opener = request.urlopen
+            reply = opener(url).read()
```
Please use `requests.get` for consistency with other uses withing qwc_services, see i.e. [1].

[1] https://github.com/qwc-services/qwc-config-generator/blob/master/config_generator/capabilities_reader.py#L250

benoitblanc commented:
Hello @manisandro , I have just pushed new commit to resolve consistency issues.

Thank you for reviewing and for your explanations.

manisandro commented:
Thanks @benoitblanc !
